File: lib/screens/redesign/redesign_overlays.dart
// Overlays drawn above the app content: the full-screen Search sheet, the
// Live Update glass tracking card, and the Toast snackbar.

import 'package:flutter/material.dart';
import 'package:material_symbols_icons/symbols.dart';

import '../../data/data_store.dart';
import '../../data/lta_models.dart';
import '../../data/mrt_geo.dart';
import '../../data/mrt_stations.dart';
import 'redesign_bridge.dart';
import 'redesign_common.dart';
import 'redesign_controller.dart';
import 'redesign_theme.dart';

// =============================================================== SEARCH

class RdSearchOverlay extends StatefulWidget {
  const RdSearchOverlay({super.key, required this.c});
  final RedesignController c;

  @override
  State<RdSearchOverlay> createState() => _RdSearchOverlayState();
}

class _RdSearchOverlayState extends State<RdSearchOverlay> {
  final TextEditingController _ctrl = TextEditingController();
  final FocusNode _focus = FocusNode();

  @override
  void initState() {
    super.initState();
    _ctrl.addListener(() => setState(() {}));
  }

  @override
  void dispose() {
    _ctrl.dispose();
    _focus.dispose();
    super.dispose();
  }

  Future<void> _openBus(String svc) async {
    final origin = await DataStore.shared.originStop(svc);
    if (mounted) widget.c.openBus(svc, origin?.busStopCode);
  }

  @override
  Widget build(BuildContext context) {
    final c = widget.c;
    final t = RdTheme.of(context);
    final q = _ctrl.text.trim();
    final services = q.isEmpty ? const <LtaBusService>[] : DataStore.shared.searchServices(q);
    final stations = q.isEmpty ? const <MrtGeoStation>[] : MrtGeo.matching(q);
    final stops = q.isEmpty ? const <LtaBusStop>[] : DataStore.shared.searchStops(q);

    final rows = <Widget>[];
    if (q.isEmpty) {
      rows.add(Padding(
        padding: const EdgeInsets.fromLTRB(40, 60, 40, 0),
        child: Column(children: [
          RdIcon(Symbols.search, size: 30, color: t.outline),
          const SizedBox(height: 8),
          Text('Search for a stop, bus number or MRT station',
              textAlign: TextAlign.center,
              style: rdText(size: 13, weight: FontWeight.w500, color: t.onVariant)),
        ]),
      ));
    } else if (services.isEmpty && stations.isEmpty && stops.isEmpty) {
      rows.add(Padding(
        padding: const EdgeInsets.only(top: 50),
        child: Center(
          child: Text('No matches for “$q”', style: rdText(size: 13, weight: FontWeight.w500, color: t.onVariant)),
        ),
      ));
    } else {
      if (services.isNotEmpty) {
        rows.add(_label(t, 'BUSES'));
        for (final s in services.take(6)) {
          rows.add(_ResultRow(
            iconBg: t.primaryContainer,
            iconColor: t.onPrimaryContainer,
            icon: Symbols.directions_bus,
            bold: 'Bus ',
            rest: s.serviceNo,
            sub: 'Tap to see the route',
            onTap: () => _openBus(s.serviceNo),
          ));
        }
      }
      if (stations.isNotEmpty) {
        rows.add(_label(t, 'MRT / LRT'));
        for (final st in stations.take(6)) {
          final code = st.codes.isNotEmpty ? st.codes.first : '';
          rows.add(_ResultRow(
            iconBg: lineColorFor(code),
            iconColor: rdMrtBadgeFg(code),
            icon: Symbols.train,
            bold: st.name,
            rest: '',
            sub: st.codes.join(' · '),
            onTap: () => c.openStationNamed(st.name),
          ));
        }
      }
      if (stops.isNotEmpty) {
        rows.add(_label(t, 'STOPS'));
        for (final s in stops.take(12)) {
          rows.add(_ResultRow(
            iconBg: t.busContainer,
            iconColor: t.onBusContainer,
            icon: Symbols.signpost,
            bold: s.description,
            rest: '',
            sub: '${s.roadName} · ${s.busStopCode}',
            onTap: () => c.openStopCode(s.busStopCode),
          ));
        }
      }
    }

    return Positioned.fill(
      child: Container(
        color: t.surface,
        child: SafeArea(
          bottom: false,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              Padding(
                padding: const EdgeInsets.fromLTRB(12, 6, 12, 10),
                child: Container(
                  height: 54,
                  padding: const EdgeInsets.symmetric(horizontal: 14),
                  decoration: BoxDecoration(color: t.scHigh, borderRadius: BorderRadius.circular(27)),
                  child: Row(
                    children: [
                      GestureDetector(onTap: c.closeSearch, child: RdIcon(Symbols.arrow_back, size: 23, color: t.onSurface)),
                      const SizedBox(width: 11),
                      Expanded(
                        child: TextField(
                          controller: _ctrl,
                          focusNode: _focus,
                          autofocus: true,
                          autocorrect: false,
                          textCapitalization: TextCapitalization.none,
                          cursorColor: t.primary,
                          style: rdText(size: 16, weight: FontWeight.w500, color: t.onSurface),
                          decoration: InputDecoration.collapsed(
                            hintText: 'Search stops, buses, MRT',
                            hintStyle: rdText(size: 16, weight: FontWeight.w500, color: t.onVariant),
                          ),
                        ),
                      ),
                      if (_ctrl.text.isNotEmpty)
                        GestureDetector(onTap: () => _ctrl.clear(), child: RdIcon(Symbols.close, size: 22, color: t.onVariant)),
                    ],
                  ),
                ),
              ),
              Expanded(
                child: ListView(padding: const EdgeInsets.symmetric(horizontal: 8), children: rows),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _label(RdTokens t, String s) => Padding(
        padding: const EdgeInsets.fromLTRB(14, 8, 14, 6),
        child: Text(s, style: rdText(size: 12, weight: FontWeight.w700, color: t.onVariant, letterSpacing: 0.24)),
      );
}

class _ResultRow extends StatelessWidget {
  const _ResultRow({
    required this.iconBg,
    required this.iconColor,
    required this.icon,
    required this.bold,
    required this.rest,
    required this.sub,
    required this.onTap,
  });
  final Color iconBg;
  final Color iconColor;
  final IconData icon;
  final String bold;
  final String rest;
  final String sub;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    return GestureDetector(
      onTap: onTap,
      behavior: HitTestBehavior.opaque,
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 10),
        child: Row(
          children: [
            Container(
              width: 40,
              height: 40,
              decoration: BoxDecoration(color: iconBg, borderRadius: BorderRadius.circular(12)),
              alignment: Alignment.center,
              child: RdIcon(icon, size: 20, color: iconColor, fill: 1),
            ),
            const SizedBox(width: 13),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text.rich(
                    TextSpan(children: [
                      TextSpan(text: bold, style: rdText(size: 15, weight: FontWeight.w700, color: t.onSurface)),
                      TextSpan(text: rest, style: rdText(size: 15, weight: FontWeight.w600, color: t.onSurface)),
                    ]),
                    maxLines: 1,
                    overflow: TextOverflow.ellipsis,
                  ),
                  Text(sub, maxLines: 1, overflow: TextOverflow.ellipsis, style: rdText(size: 12, weight: FontWeight.w500, color: t.onVariant)),
                ],
              ),
            ),
            RdIcon(Symbols.north_west, size: 16, color: t.outline),
          ],
        ),
      ),
    );
  }
}

// ============================================================ LIVE UPDATE

class RdLiveUpdate extends StatelessWidget {
  const RdLiveUpdate({super.key, required this.c});
  final RedesignController c;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      top: 8,
      left: 10,
      right: 10,
      child: SafeArea(
        bottom: false,
        child: _LiveCard(c: c),
      ),
    );
  }
}

class _LiveCard extends StatelessWidget {
  const _LiveCard({required this.c});
  final RedesignController c;

  static const _accent = Color(0xFF9CC0FF);
  static const _onAccent = Color(0xFF10245E);
  static const _white = Color(0xFFFFFFFF);

  String get _svc => c.activeService ?? '';
  String get _trackedStop =>
      c.activeRouteStop != null ? DataStore.shared.stopName(c.activeRouteStop!) : 'your stop';
  String get _eta {
    final code = c.activeRouteStop;
    if (code == null) return '—';
    for (final s in DataStore.shared.servicesFor(code)) {
      if (s.no == _svc) return rdMinLabel(s.etaSec);
    }
    return '—';
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 15),
      decoration: BoxDecoration(
        color: const Color(0xD114121C),
        borderRadius: BorderRadius.circular(26),
        border: Border.all(color: const Color(0x1FFFFFFF)),
        boxShadow: const [BoxShadow(color: Color(0x80000000), blurRadius: 40, offset: Offset(0, 16))],
      ),
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            children: [
              Container(
                width: 24,
                height: 24,
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(7),
                  gradient: const LinearGradient(
                    begin: Alignment.topLeft,
                    end: Alignment.bottomRight,
                    colors: [Color(0xFF2C72E6), Color(0xFF222A38)],
                  ),
                ),
                alignment: Alignment.center,
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    RdIcon(Symbols.directions_bus, size: 8, color: _white, fill: 1),
                    RdIcon(Symbols.train, size: 8, color: _white, fill: 1),
                  ],
                ),
              ),
              const SizedBox(width: 9),
              Text('SG Transit', style: rdText(size: 12, weight: FontWeight.w700, color: const Color(0xE6FFFFFF))),
              const SizedBox(width: 8),
              Container(
                padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 2),
                decoration: BoxDecoration(color: const Color(0x389CC0FF), borderRadius: BorderRadius.circular(8)),
                child: Row(
                  mainAxisSize: MainAxisSize.min,
                  children: [
                    const RdDot(_accent, size: 5),
                    const SizedBox(width: 4),
                    Text('LIVE', style: rdText(size: 10, weight: FontWeight.w800, color: _accent)),
                  ],
                ),
              ),
              const Spacer(),
              GestureDetector(
                onTap: c.dismissLU,
                child: RdIcon(Symbols.close, size: 18, color: const Color(0x8CFFFFFF)),
              ),
            ],
          ),
          const SizedBox(height: 12),
          Row(
            crossAxisAlignment: CrossAxisAlignment.end,
            children: [
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text('Bus $_svc → your stop', style: rdText(size: 16, weight: FontWeight.w700, color: _white)),
                    const SizedBox(height: 2),
                    Text('Approaching $_trackedStop',
                        maxLines: 1,
                        overflow: TextOverflow.ellipsis,
                        style: rdText(size: 12.5, weight: FontWeight.w500, color: const Color(0xA6FFFFFF))),
                  ],
                ),
              ),
              Column(
                crossAxisAlignment: CrossAxisAlignment.end,
                children: [
                  Text(_eta, style: rdText(size: 36, weight: FontWeight.w900, color: _accent, height: 1, letterSpacing: -1.08)),
                  Text('min', style: rdText(size: 11, weight: FontWeight.w600, color: const Color(0xB3FFFFFF))),
                ],
              ),
            ],
          ),
          const SizedBox(height: 12),
          // progress bar at 78%
          LayoutBuilder(builder: (context, cons) {
            return SizedBox(
              height: 18,
              child: Stack(
                clipBehavior: Clip.none,
                alignment: Alignment.centerLeft,
                children: [
                  Container(
                    height: 6,
                    decoration: BoxDecoration(color: const Color(0x26FFFFFF), borderRadius: BorderRadius.circular(3)),
                  ),
                  FractionallySizedBox(
                    widthFactor: 0.78,
                    child: Container(
                      height: 6,
                      decoration: BoxDecoration(color: _accent, borderRadius: BorderRadius.circular(3)),
                    ),
                  ),
                  Positioned(
                    left: cons.maxWidth * 0.78 - 9,
                    child: Container(
                      width: 18,
                      height: 18,
                      decoration: BoxDecoration(
                        color: _accent,
                        borderRadius: BorderRadius.circular(6),
                        boxShadow: const [BoxShadow(color: Color(0x409CC0FF), blurRadius: 0, spreadRadius: 4)],
                      ),
                      alignment: Alignment.center,
                      child: RdIcon(Symbols.directions_bus, size: 11, color: _onAccent, fill: 1),
                    ),
                  ),
                ],
              ),
            );
          }),
          const SizedBox(height: 14),
          Row(
            children: [
              Expanded(
                child: GestureDetector(
                  onTap: c.stopTrack,
                  child: Container(
                    height: 42,
                    decoration: BoxDecoration(
                      borderRadius: BorderRadius.circular(21),
                      border: Border.all(color: const Color(0x47FFFFFF)),
                    ),
                    alignment: Alignment.center,
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        RdIcon(Symbols.stop_circle, size: 18, color: _white),
                        const SizedBox(width: 7),
                        Text('Stop', style: rdText(size: 13.5, weight: FontWeight.w700, color: _white)),
                      ],
                    ),
                  ),
                ),
              ),
              const SizedBox(width: 9),
              Expanded(
                child: GestureDetector(
                  onTap: c.luView,
                  child: Container(
                    height: 42,
                    decoration: BoxDecoration(color: _accent, borderRadius: BorderRadius.circular(21)),
                    alignment: Alignment.center,
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        RdIcon(Symbols.map, size: 18, color: _onAccent),
                        const SizedBox(width: 7),
                        Text('View route', style: rdText(size: 13.5, weight: FontWeight.w700, color: _onAccent)),
                      ],
                    ),
                  ),
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }
}

// ================================================================= TOAST

class RdToast extends StatelessWidget {
  const RdToast({super.key, required this.c});
  final RedesignController c;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      left: 12,
      right: 12,
      bottom: 14,
      child: SafeArea(
        top: false,
        child: Container(
          padding: const EdgeInsets.fromLTRB(15, 13, 14, 13),
          decoration: BoxDecoration(
            color: const Color(0xFF2C2A33),
            borderRadius: BorderRadius.circular(16),
            boxShadow: const [BoxShadow(color: Color(0x8C000000), blurRadius: 34, offset: Offset(0, 14))],
          ),
          child: Row(
            children: [
              RdIcon(Symbols.notifications_active, size: 21, color: const Color(0xFF9CC0FF), fill: 1),
              const SizedBox(width: 12),
              Expanded(
                child: Text(c.toast ?? '',
                    style: rdText(size: 12.5, weight: FontWeight.w500, color: const Color(0xFFE9E5EE), height: 1.35)),
              ),
              const SizedBox(width: 12),
              GestureDetector(
                onTap: c.dismissToast,
                child: Text('Got it', style: rdText(size: 13, weight: FontWeight.w700, color: const Color(0xFF9CC0FF))),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

File: ios-native/Leyne/Redesign/RedesignOverlays.swift
// Overlays drawn above the app content (iOS): the full-screen Search sheet, the
// Live Update glass tracking card, and the Toast snackbar.

import SwiftUI
import MapKit

// =================================================================== SEARCH

struct RDSearchOverlay: View {
    @ObservedObject var m: RedesignModel
    let t: RDTokens
    @EnvironmentObject private var store: DataStore
    @State private var query = ""
    // Building/place hits (MapKit) + a "focused location" (a tapped place or a
    // geocoded postal code) whose nearby stops/stations we then list.
    @State private var places: [MKMapItem] = []
    @State private var focusTitle: String? = nil
    @State private var focusLat: Double = 0
    @State private var focusLon: Double = 0
    @FocusState private var focused: Bool

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 11) {
                Button(action: { m.closeSearch() }) { RDSym("arrow.left", size: 23, color: t.onSurface) }
                    .buttonStyle(.plain)
                TextField("Stops · buses · MRT · postal · places", text: $query)
                    .focused($focused)
                    .font(rdFont(16, .medium))
                    .foregroundStyle(t.onSurface)
                    .autocorrectionDisabled()
                    .textInputAutocapitalization(.never)
                    .submitLabel(.search)
                if !query.isEmpty {
                    Button(action: { query = ""; places = []; focusTitle = nil }) {
                        RDSym("xmark", size: 22, color: t.onVariant)
                    }
                    .buttonStyle(.plain)
                }
            }
            .padding(.horizontal, 14).frame(height: 54)
            .background(t.scHigh).clipShape(Capsule())
            .padding(.horizontal, 12).padding(.top, 6).padding(.bottom, 10)

            ScrollView {
                VStack(spacing: 0) {
                    let q = query.trimmingCharacters(in: .whitespaces)
                    if q.isEmpty { hint } else { results(q) }
                }
                .padding(.horizontal, 8)
            }
        }
        .background(t.surface.ignoresSafeArea())
        .onAppear { focused = true }
        .task(id: query) { await runSearch() }
    }

    /// A 6-digit postal code geocodes straight to a focused location; anything
    /// else runs a MapKit place/building search (biased to Singapore). Cancels
    /// automatically as the query changes (via `.task(id:)`), so it debounces.
    private func runSearch() async {
        let q = query.trimmingCharacters(in: .whitespaces)
        focusTitle = nil
        guard q.count >= 3 else { places = []; return }
        if q.count == 6, q.allSatisfy(\.isNumber) {
            places = []
            if let c = try? await CLGeocoder().geocodeAddressString("Singapore \(q)").first?.location?.coordinate {
                focusTitle = "Postal \(q)"; focusLat = c.latitude; focusLon = c.longitude
            }
            return
        }
        let req = MKLocalSearch.Request()
        req.naturalLanguageQuery = q
        req.region = MKCoordinateRegion(center: CLLocationCoordinate2D(latitude: 1.3521, longitude: 103.8198),
                                        span: MKCoordinateSpan(latitudeDelta: 0.7, longitudeDelta: 0.7))
        if let resp = try? await MKLocalSearch(request: req).start() {
            places = Array(resp.mapItems.prefix(6))
        } else {
            places = []
        }
    }

    /// Bus stops within ~700 m of a coordinate, nearest first.
    private func stopsNear(_ lat: Double, _ lon: Double, limit: Int = 8) -> [LTABusStop] {
        store.stopByCode.values
            .map { ($0, haversine(lat, lon, $0.Latitude, $0.Longitude)) }
            .filter { $0.1 <= 700 }
            .sorted { $0.1 < $1.1 }
            .prefix(limit).map { $0.0 }
    }

    private var hint: some View {
        VStack(spacing: 8) {
            RDSym("magnifyingglass", size: 30, color: t.outline)
            Text("Search for a stop, bus number or MRT station")
                .font(rdFont(13, .medium)).foregroundStyle(t.onVariant).multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity).padding(.top, 60).padding(.horizontal, 40)
    }

    @ViewBuilder private func results(_ q: String) -> some View {
        if let title = focusTitle {
            nearResults(title: title, lat: focusLat, lon: focusLon)
        } else {
            let services = store.searchServices(q)
            let stations = MrtGeo.stations(matching: q)
            let stops = store.searchStops(q)
            if services.isEmpty && stations.isEmpty && stops.isEmpty && places.isEmpty {
                Text("No matches for “\(q)”")
                    .font(rdFont(13, .medium)).foregroundStyle(t.onVariant)
                    .frame(maxWidth: .infinity).padding(.top, 50)
            } else {
                if !services.isEmpty {
                    label("BUSES")
                    ForEach(services.prefix(6), id: \.ServiceNo) { svc in
                        resultRow(t.primaryContainer, t.onPrimaryContainer, "bus.fill",
                                  "Bus ", svc.ServiceNo, "Tap to see the route") { openBus(svc.ServiceNo) }
                    }
                }
                if !stations.isEmpty {
                    label("MRT / LRT")
                    ForEach(Array(stations.prefix(6))) { st in
                        let code = st.codes.first ?? ""
                        resultRow(mrtLineColorFor(code), rdMrtBadgeFg(code), "tram.fill",
                                  st.name, "", st.codes.joined(separator: " · ")) { m.openStation(named: st.name) }
                    }
                }
                if !stops.isEmpty {
                    label("STOPS")
                    ForEach(stops.prefix(10), id: \.BusStopCode) { s in
                        resultRow(t.busContainer, t.onBusContainer, "signpost.right.fill",
                                  s.Description, "", "\(s.RoadName) · \(s.BusStopCode)") { m.openStop(code: s.BusStopCode) }
                    }
                }
                if !places.isEmpty {
                    label("PLACES")
                    ForEach(places, id: \.self) { item in placeRow(item) }
                }
            }
        }
    }

    /// A building / place hit — tapping it drills into the stops near it.
    private func placeRow(_ item: MKMapItem) -> some View {
        let name = item.name ?? "Place"
        let c = item.placemark.coordinate
        return resultRow(t.amberContainer, t.onAmberContainer, "building.2.fill",
                         name, "", item.placemark.title ?? "See nearby stops") {
            focusTitle = name; focusLat = c.latitude; focusLon = c.longitude
        }
    }

    /// "Near <place/postal>" — the bus stops and MRT stations around a focused
    /// location (a tapped place or a geocoded postal code).
    @ViewBuilder private func nearResults(title: String, lat: Double, lon: Double) -> some View {
        Button(action: { focusTitle = nil }) {
            HStack(spacing: 6) {
                RDSym("chevron.left", size: 15, color: t.primary)
                Text("Near \(title)").font(rdFont(13, .bold)).foregroundStyle(t.primary).lineLimit(1)
                Spacer()
            }
            .padding(.horizontal, 14).padding(.vertical, 9)
        }
        .buttonStyle(.plain)
        let stops = stopsNear(lat, lon)
        let stations = MrtGeo.nearestStations(
            to: CLLocationCoordinate2D(latitude: lat, longitude: lon), limit: 3, withinMeters: 1000)
        if stops.isEmpty && stations.isEmpty {
            Text("No stops within walking distance")
                .font(rdFont(13, .medium)).foregroundStyle(t.onVariant)
                .frame(maxWidth: .infinity).padding(.top, 40)
        } else {
            if !stops.isEmpty {
                label("BUS STOPS")
                ForEach(stops, id: \.BusStopCode) { s in
                    let d = Int(haversine(lat, lon, s.Latitude, s.Longitude).rounded())
                    resultRow(t.busContainer, t.onBusContainer, "signpost.right.fill",
                              s.Description, "", "\(max(1, d / 80)) min walk · \(s.RoadName)") { m.openStop(code: s.BusStopCode) }
                }
            }
            if !stations.isEmpty {
                label("MRT / LRT")
                ForEach(stations, id: \.station.id) { item in
                    let st = item.station; let code = st.codes.first ?? ""
                    resultRow(mrtLineColorFor(code), rdMrtBadgeFg(code), "tram.fill",
                              st.name, "", "\(item.walkMin) min walk") { m.openStation(named: st.name) }
                }
            }
        }
    }

    private func openBus(_ svc: String) {
        Task {
            let origin = await store.originStop(ofService: svc)
            await MainActor.run { m.openBus(service: svc, stopCode: origin?.BusStopCode) }
        }
    }

    private func label(_ s: String) -> some View {
        Text(s).font(rdFont(12, .bold)).foregroundStyle(t.onVariant)
            .frame(maxWidth: .infinity, alignment: .leading).padding(.horizontal, 14).padding(.top, 8).padding(.bottom, 6)
    }

    private func resultRow(_ iconBg: Color, _ iconColor: Color, _ symbol: String,
                           _ bold: String, _ rest: String, _ sub: String,
                           _ tap: @escaping () -> Void) -> some View {
        Button(action: tap) {
            HStack(spacing: 13) {
                RDSym(symbol, size: 20, color: iconColor)
                    .frame(width: 40, height: 40)
                    .background(iconBg).clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
                VStack(alignment: .leading, spacing: 0) {
                    (Text(bold).font(rdFont(15, .bold)) + Text(rest).font(rdFont(15, .semibold)))
                        .foregroundStyle(t.onSurface).lineLimit(1)
                    Text(sub).font(rdFont(12, .medium)).foregroundStyle(t.onVariant).lineLimit(1)
                }
                Spacer()
                RDSym("arrow.up.left", size: 16, color: t.outline)
            }
            .padding(.horizontal, 14).padding(.vertical, 10)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }
}

// ============================================================== LIVE UPDATE

struct RDLiveUpdate: View {
    @ObservedObject var m: RedesignModel
    let t: RDTokens
    @EnvironmentObject private var store: DataStore
    @State private var live = true

    private let accent = rdHex("9CC0FF")
    private let onAccent = rdHex("10245E")

    private var svc: String { m.activeService ?? "" }
    private var trackedStop: String { m.activeRouteStop.map { store.stopName($0) } ?? "your stop" }
    private var etaLabel: String {
        guard let code = m.activeRouteStop,
              let s = store.servicesFor(code).first(where: { $0.no == svc }) else { return "—" }
        return rdMinLabel(s.etaSec)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 8) {
                RoundedRectangle(cornerRadius: 7, style: .continuous)
                    .fill(LinearGradient(colors: [rdHex("2C72E6"), rdHex("222A38")], startPoint: .topLeading, endPoint: .bottomTrailing))
                    .frame(width: 24, height: 24)
                    .overlay(HStack(spacing: 0) {
                        RDSym("bus.fill", size: 8, color: .white)
                        RDSym("tram.fill", size: 8, color: .white)
                    })
                Text("SG Transit").font(rdFont(12, .bold)).foregroundStyle(.white.opacity(0.9))
                HStack(spacing: 4) {
                    RDDot(color: accent, size: 5).opacity(live ? 1 : 0.3)
                    Text("LIVE").font(rdFont(10, .heavy)).foregroundStyle(accent)
                }
                .padding(.horizontal, 8).padding(.vertical, 2)
                .background(accent.opacity(0.22)).clipShape(RoundedRectangle(cornerRadius: 8, style: .continuous))
                Spacer()
                Button(action: { m.dismissLU() }) { RDSym("xmark", size: 18, color: .white.opacity(0.55)) }
                    .buttonStyle(.plain)
            }
            .padding(.bottom, 12)

            HStack(alignment: .bottom) {
                VStack(alignment: .leading, spacing: 2) {
                    Text("Bus \(svc) → your stop").font(rdFont(16, .bold)).foregroundStyle(.white)
                    Text("Approaching \(trackedStop)").font(rdFont(12.5, .medium)).foregroundStyle(.white.opacity(0.65)).lineLimit(1)
                }
                Spacer()
                VStack(alignment: .trailing, spacing: 0) {
                    Text(etaLabel).font(rdFont(36, .black)).foregroundStyle(accent)
                    Text("min").font(rdFont(11, .semibold)).foregroundStyle(.white.opacity(0.7))
                }
            }
            .padding(.bottom, 12)

            GeometryReader { geo in
                ZStack(alignment: .leading) {
                    Capsule().fill(Color.white.opacity(0.15)).frame(height: 6)
                    Capsule().fill(accent).frame(width: geo.size.width * 0.78, height: 6)
                    RoundedRectangle(cornerRadius: 6, style: .continuous).fill(accent)
                        .frame(width: 18, height: 18)
                        .overlay(RDSym("bus.fill", size: 11, color: onAccent))
                        .offset(x: geo.size.width * 0.78 - 9)
                }
            }
            .frame(height: 18)
            .padding(.bottom, 14)

            HStack(spacing: 9) {
                Button(action: { m.stopTrack() }) {
                    HStack(spacing: 7) {
                        RDSym("stop.circle", size: 18, color: .white)
                        Text("Stop").font(rdFont(13.5, .bold)).foregroundStyle(.white)
                    }
                    .frame(maxWidth: .infinity).frame(height: 42)
                    .overlay(Capsule().strokeBorder(.white.opacity(0.28), lineWidth: 1))
                }.buttonStyle(.plain)
                Button(action: { m.luView() }) {
                    HStack(spacing: 7) {
                        RDSym("map.fill", size: 18, color: onAccent)
                        Text("View route").font(rdFont(13.5, .bold)).foregroundStyle(onAccent)
                    }
                    .frame(maxWidth: .infinity).frame(height: 42)
                    .background(accent).clipShape(Capsule())
                }.buttonStyle(.plain)
            }
        }
        .padding(.horizontal, 16).padding(.vertical, 15)
        .background(rdHex("14121C").opacity(0.92))
        .clipShape(RoundedRectangle(cornerRadius: 26, style: .continuous))
        .overlay(RoundedRectangle(cornerRadius: 26, style: .continuous).strokeBorder(.white.opacity(0.12), lineWidth: 1))
        .shadow(color: .black.opacity(0.5), radius: 24, x: 0, y: 16)
        .padding(.horizontal, 10).padding(.top, 8)
        .onAppear { withAnimation(.easeInOut(duration: 1.4).repeatForever()) { live = false } }
    }
}

// ==================================================================== TOAST

struct RDToast: View {
    @ObservedObject var m: RedesignModel
    let t: RDTokens

    var body: some View {
        HStack(spacing: 12) {
            RDSym("bell.fill", size: 21, color: rdHex("9CC0FF"))
            Text(m.toast ?? "").font(rdFont(12.5, .medium)).foregroundStyle(rdHex("E9E5EE"))
                .frame(maxWidth: .infinity, alignment: .leading)
            Button(action: { m.dismissToast() }) {
                Text("Got it").font(rdFont(13, .bold)).foregroundStyle(rdHex("9CC0FF"))
            }.buttonStyle(.plain)
        }
        .padding(.leading, 15).padding(.trailing, 14).padding(.vertical, 13)
        .background(rdHex("2C2A33"))
        .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
        .shadow(color: .black.opacity(0.55), radius: 17, x: 0, y: 14)
        .padding(.horizontal, 12)
    }
}
